### DISCUSSION TITLE
How to find a maximum-weight matching in a bipartite graph?

### DISCUSSION THREAD

**Discussion by erelsgl** [networkx/networkx#4683] [Q&A]
I found in the docs, an algorithm for finding a maximum-weight matching in a general graph, and for finding a maximum-cardinality matching in a bipartite graph.  Is there an algorithm for finding a maximum-weight matching in a bipartite graph? It could run much faster than the algorithm for general graphs.

**Comment by dschult** [comment]
There is `minimum_weight_full_matching` for bipartite graphs. So maximum weight could be obtained by negating the weights.
It might be easy to write a function that does that conversion and calls the existing function. If that function is not easy, or not easy to generalize, we should include such a function in `bipartite/matching.py`.

Does this work? (untested!)
````python
weight_dict = nx.get_edge_attribute(G, "weight")
nx.set_edge_attribute(G, {e:-wt for e, wt in weight_dict.items()}, "new_weight")
match = minimum_weight_full_matching(G, top_nodes=T, "new_weight")
```

**Reply by erelsgl** [reply]
Even if it is correct, it would give you maximum-weight full matching, which is useful in itself, but it is not the same as maximum weight matching. For example, consider a bipartite graph with the following edges:

* a--x: weight 1
* b--y: weight 1
* a--y: weight 8

The max-weight matching is a--y, but the max-weight full matching is (a--x, b--y).

**Comment by dschult** [comment]
Ahh...   Thanks!
That's an important distinction. 

We don't have algorithms to find maximum weighted matchings for bipartite graphs.

**Comment by fuglede** [comment]
Note that a simple reduction from maximum weight perfect matching to maximum weight matching is given in [Gabow, Tarjan - *Faster scaling algorithms for network problems*](http://www.columbia.edu/~cs2035/courses/ieor8100.F12/GabTar.pdf), Sect. 2.2, paraphrased:

> Form G by taking two copies of the bipartite graph; for each vertex v in the original graph join the two copies of v by a cost zero edge. Then G is bipartite, and a minimum perfect matching in G gives a minimum-cost matching in the original graph.

With the example you give above (using the underlying `scipy.optimize.linear_sum_assignment` for simplicity):

```python
# Create the doubled graph
a = np.array([[1, 8], [-np.inf, 1]])
cost = np.block([[a, np.zeros((a.shape[0], a.shape[0]))],
                 [np.zeros((a.shape[1], a.shape[1])), a.T]])
row_ind, col_ind = linear_sum_assignment(cost, maximize=True)

# Pull out the matching in the original graph
mask = (row_ind < a.shape[0]) & (col_ind < a.shape[1])
row_ind = row_ind[mask]
col_ind = col_ind[mask]

# Let's check the total cost
print(cost[row_ind, col_ind].sum())  # 8.0
```

Note that similar reductions are discussed in https://github.com/scipy/scipy/issues/14041
